Command bar refactor (Ready for review/merge to 5.0)

micahgodbolt opened the pull request.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.Props.ts:
```diff
@@ -16,39 +20,74 @@ export interface ICommandBarProps extends React.HTMLAttributes<HTMLDivElement> {
+   * Context under which the item is being rendered
+   * This value is controlled by the component and useful for adjusting onRender function
+   */
+  renderedInOverflow?: boolean;
```
Have you considered instead having two separate onRender functions? One for the top level and one for the overflow.

micahgodbolt replied:
That'd need to be supported at the overflowSet layer ( i believe). I could go either way, I just know we'll want a way to differentiate.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx:
```diff
@@ -1,427 +1,187 @@
+  primaryItems: ICommandBarItemProps[];
+  overflowItems: ICommandBarItemProps[];
+  farItems: ICommandBarItemProps[] | undefined;
+  originalOverflowItems: ICommandBarItemProps[];
```
Comments on the properties please.

christiango replied:
What is originalOverflowItems? Is that the default contents of the overflow menu?

micahgodbolt replied:
yeah, if I say "put these 3 in the overflow to start", we don't want to accidentally pull them out when onGrowData.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx:
```diff
@@ -1,427 +1,187 @@
-          ) : (null)
-        }
-      </div >
+      <ResizeGroup
```
How is this working with ResizeGroup if the items are display flex and there is a left and a right group? Doesn't the measured width of the contents in the hidden div depend on the container width?

christiango replied:
Looking later at the code, it appears only the first group resizes. Should that be the only thing in the overflow set then? (Also see my comment about simplifying the component to not have far commands.

micahgodbolt replied:
I'm happy to chat about possible approaches, but all of our command bar designs include those far items. And often the right items change, reducing the space for the left items, pushing items into the overflow. It seemed logical to put them together to allow resizeGroup to look at the entire row of buttons.

christiango replied:
Is the measurement being cached accurate then? When rendering in the hidden div, won't it put a space between the elements?

christiango replied:
Also, this approach means the component is bloated for people who don't care about the far controls.

micahgodbolt replied:
It renders the entire left/right sides and measures the width needed for both, and then reduces the left side if there isn't room. This is the reason I render both sides in this component so that the measurement can be made with all of the necessary content.

As for bloat, i'm not introducing any new components to support the right hand side, and it's a couple lines of code to render/position them.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx:
```diff
@@ -1,427 +1,187 @@
-      this._updateRenderedItems();
-    });
-  }
+      overflowItems = overflowItems.concat(movedItem);
```
I personally prefer using array spread over concat, not sure what the fabric convention is

micahgodbolt replied:
spreads are beautiful :) fixed

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx:
```diff
@@ -1,427 +1,187 @@
-      })() }
-    </div>;
+  private computeCacheKey(primaryItems: ICommandBarItemProps[]): string {
+    return primaryItems.reduce((acc, current) => acc + current.key, '');
```
This probably also needs to take into account if there is an overflow well.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx:
```diff
@@ -1,427 +1,187 @@
+              />
+
+              {/*Secondary Items*/ }
+              <OverflowSet
```
So the far commands never render in an overflow. I'm wondering if this makes sense to be taken out of command bar and leaving command bar to just deal with the part that resizes? Consumers could still implement a far commands by composing multiple components

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx:
```diff
@@ -1,427 +1,187 @@
-        }
-      })() }
-    </div>;
+  private computeCacheKey(primaryItems: ICommandBarItemProps[]): string {
```
We also need to be very careful here, we should make it clear on the props comments that things with the same key are assumed to have the same horizontal width. We need to make sure people simply aren't assigning indices in an array or something like that since that will be buggy.

Do we have a scenario where a command sometimes appears in the top bar with a label, but after some scaling logic it no longer has a label? If so, we need to make sure the label and no label cases have different cache keys

micahgodbolt replied:
yeah, we'd need to figure out a way to invalidate the key if the width changed.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.tsx:
```diff
@@ -1,427 +1,187 @@
-      } else {
-        // Find the currently open key in the new props
-        let matchingItem = renderedItems.filter(item => item.key === this.state.expandedMenuItemKey);
+    const commandButtonProps: ICommandBarItemProps = assign({}, item, {
```
object spread?

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBarOld.old:
```diff
@@ -0,0 +1,410 @@
+  BaseComponent,
+  autobind,
+  buttonProperties,
+  anchorProperties,
```
remove this file?

micahgodbolt replied:
yeah, was just keeping it around for reference.

christiango commented:
This is also missing some unit tests. It might be helpful to extract some of the logic for computing the next overflow items into functions that could be exported and tested in isolation.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.scss:
```diff
@@ -1,317 +1,43 @@
-
-.primaryCommands {
-  overflow: hidden;
+  .search {
```
The baked in search feels broken. It results in unecessary bytes being pulled down for people that don't use a search box.

micahgodbolt replied:
oh yeah, search came out, that was just leftover styles I forgot to remove

phkuo commented on packages/office-ui-fabric-react/src/components/Button/SplitButton/SplitButton.classNames.ts:
```diff
@@ -0,0 +1,33 @@
+  return {
+    root: mergeStyles(
+      styles.splitButtonMenuButton,
+      expanded && [
```
I'd leave the array off for legibility if you don't need it, personal preference though.

phkuo commented on packages/styling/src/styles/UtilityStyles.ts:
```diff
@@ -0,0 +1,12 @@
+
+export const hideText: IRawStyle = {
+  position: 'absolute',
+  width: '1px',
```
display:block so the dimensions actually apply

phkuo commented on packages/styling/src/styles/UtilityStyles.ts:
```diff
@@ -0,0 +1,12 @@
+  position: 'absolute',
+  width: '1px',
+  height: '1px',
+  margin: '-1px',
```
pretty sure margin doesn't do anything when you're position:Absolute?

micahgodbolt replied:
drat...this code shouldn't be in this PR. As for that line, i just moved it, didn't write it :)

phkuo commented on packages/styling/src/styles/UtilityStyles.ts:
```diff
@@ -0,0 +1,12 @@
+import { IRawStyle } from '../interfaces/index';
+
+export const hideText: IRawStyle = {
+  position: 'absolute',
```
also top:-9999px and left:-9999px

phkuo commented on packages/styling/src/styles/UtilityStyles.ts:
```diff
@@ -0,0 +1,12 @@
+import { IRawStyle } from '../interfaces/index';
+
+export const hideText: IRawStyle = {
```
I would call this "accessibleText" since that's the purpose of it.

micahgodbolt commented:
@christiango I've addressed your code concerns. I could certainly add more tests but this puts it on par with the old test coverage. As far as maintaining the far items, I feel the measuring logic works well, and the component overhead is so minimal that it is worth supporting what in odsp is a very common scenario.

christiango commented:
I think the old test coverage was lacking. For example, I don't think there's any coverage on the logic to move controls into the overflow menu.

christiango commented:
We don't need to test the measuring logic, but having coverage for onGrowData and onReduce data would be good, I think.

christiango commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.Props.ts:
```diff
@@ -16,39 +20,74 @@ export interface ICommandBarProps extends React.HTMLAttributes<HTMLDivElement> {
\ No newline at end of file
+}
+
+export interface ICommandBarItemProps extends IContextualMenuItem {
```
We might want to explicitly have the cacheKey as a property here and use that instead of the the key for computing the overall cache key. The semantically mean different things, key is just a render hint for the React reconciler while the cacheKey means that 'if two items have the same cacheKey, the have the same width'. An example where the difference is important would be if a consumer changes the label of a button at runtime from 'search' to 'find'. It's reasonable to expect the keys to be the same, but the cacheKeys should be different, since the width of a button with the text 'search' is different than that with 'find'.

christiango replied:
We should also only use cacheKeys when the consumer has explicitly provided them. Unless you think we can compute them accurately in fabric.

micahgodbolt replied:
key's are required, so i'll default to the key if a cacheKey is not provided.

kristoferbrown commented on packages/office-ui-fabric-react/src/components/CommandBar/CommandBar.Props.ts:
```diff
@@ -16,39 +20,79 @@ export interface ICommandBarProps extends React.HTMLAttributes<HTMLDivElement> {
 
   /**
-   * Default items to have in the overflow menu
+  * Text to be read by screen readers if there are overflow items and focus is on elipsis button
```
This comment doesn't describe the following prop.

kristoferbrown commented on packages/office-ui-fabric-react/src/components/CommandBar/examples/CommandBar.Example.scss:
```diff
@@ -1,48 +1 @@
-  background: #d0d0d0
-}
\ No newline at end of file
+@import '../../../common/common';
```
Is this file still even necessary?

micahgodbolt replied:
Nope!

battletoilet commented:
@micahgodbolt any progress here?

micahgodbolt commented:
Going to close this up and pursue getting the code into experiments first.
